### PR TITLE
Add a simple feature to recommend people use the new C# 9 'not' pattern.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/46699
Fixes https://github.com/dotnet/roslyn/issues/46798
Fixes https://github.com/dotnet/roslyn/issues/46706

Looks like:

image

### CODE REVIEW

File: src/Analyzers/CSharp/Analyzers/UsePatternMatching/CSharpUseNotPattermDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+            if (!(node is PrefixUnaryExpressionSyntax
+            {
+                Operand: ParenthesizedExpressionSyntax
+                {
+                    Expression: IsPatternExpressionSyntax
+                    {
+                        Pattern: DeclarationPatternSyntax declarationPattern,
+                    } isPattern,
+                } parenthesizedExpression,
+            } notExpression))
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UsePatternMatching/CSharpUseNotPattermDiagnosticAnalyzer.cs:73]: i don't love the idnentation here.  but until we get Sam's formatting PR flowing into us, this is required

File: src/Analyzers/CSharp/CodeFixes/UsePatternMatching/CSharpUseNotPatternCodeFixProvider.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
```
Comment (davidwengier) [inline src/Analyzers/CSharp/CodeFixes/UsePatternMatching/CSharpUseNotPatternCodeFixProvider.cs:4]: #nullable enable?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/CodeStyle/CSharpCodeStyleOptions.cs [hunk 3]
```diff
@@ -79,6 +79,13 @@ private static Option2<T> CreateOption<T>(OptionGroup group, string name, T defa
                 EditorConfigStorageLocation.ForBoolCodeStyleOption("csharp_style_pattern_matching_over_is_with_cast_check"),
                 new RoamingProfileStorageLocation($"TextEditor.CSharp.Specific.{nameof(PreferPatternMatchingOverIsWithCastCheck)}")});
 
+        public static readonly Option2<CodeStyleOption2<bool>> PreferNotPattern = CreateOption(
+            CSharpCodeStyleOptionGroups.PatternMatching, nameof(PreferNotPattern),
+            defaultValue: s_trueWithSuggestionEnforcement,
+            storageLocations: new OptionStorageLocation2[] {
+                EditorConfigStorageLocation.ForBoolCodeStyleOption("csharp_style_not_pattern"),
```
Comment (davidwengier) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/CodeStyle/CSharpCodeStyleOptions.cs:86]: It's surprising to me that editor config toggles are buried in here. Should probably add this to the docs page at least though.

File: src/Analyzers/CSharp/CodeFixes/UsePatternMatching/CSharpUseNotPatternCodeFixProvider.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/CodeFixes/UsePatternMatching/CSharpUseNotPatternCodeFixProvider.cs:4]: done.

**Comment by CyrusNajmabadi** [comment]
VB side done as well.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/CodeStyle/VisualBasicCodeStyleOptions.vb [hunk 4]
```diff
@@ -30,12 +30,18 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.CodeStyle
...
             VisualBasicCodeStyleOptionGroups.Modifier, NameOf(PreferredModifierOrder),
             defaultValue:=New CodeStyleOption2(Of String)(String.Join(",", PreferredModifierOrderDefault.Select(AddressOf SyntaxFacts.GetText)), NotificationOption2.Silent),
             EditorConfigStorageLocation.ForStringCodeStyleOption("visual_basic_preferred_modifier_order"),
             New RoamingProfileStorageLocation($"TextEditor.%LANGUAGE%.Specific.{NameOf(PreferredModifierOrder)}"))
 
+        Public Shared ReadOnly PreferIsNotExpression As Option2(Of CodeStyleOption2(Of Boolean)) = CreateOption(
+            VisualBasicCodeStyleOptionGroups.ExpressionLevelPreferences, NameOf(PreferIsNotExpression),
+            defaultValue:=New CodeStyleOption2(Of Boolean)(True, NotificationOption2.Suggestion),
+            EditorConfigStorageLocation.ForBoolCodeStyleOption("visual_basic_style_prefer_isnot_expression"),
+            New RoamingProfileStorageLocation($"TextEditor.%LANGUAGE%.Specific.{NameOf(PreferIsNotExpression)}"))
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/VisualBasic/CodeStyle/VisualBasicCodeStyleOptions.vb:43]: 📝 These are causing a pretty big problem for .editorconfig in the code style layer. I would prefer to implement this without a new code style option.

/cc @mavasani

**Comment by sharwell** [comment]
I removed the auto-merge label so we can discuss this tomorrow.

File: src/Analyzers/CSharp/CodeFixes/UsePatternMatching/CSharpUseNotPatternCodeFixProvider.cs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            return Task.CompletedTask;
+        }
+
+        private static void ProcessDiagnostic(
+            SyntaxEditor editor,
+            Diagnostic diagnostic,
+            CancellationToken cancellationToken)
+        {
+#if CODE_STYLE
+            Contract.Fail("We should have never gotten here as CODE_STYLE doesn't support C# 9 yet.");
```
Comment (sharwell) [inline src/Analyzers/CSharp/CodeFixes/UsePatternMatching/CSharpUseNotPatternCodeFixProvider.cs:67]: 📝 I'm fairly certain this _can_ get hit in the code style layer, which would be a pretty big break for the package.

File: src/Analyzers/CSharp/Analyzers/CSharpAnalyzers.projitems [hunk 6]
```diff
@@ -96,11 +96,12 @@
     <Compile Include="$(MSBuildThisFileDirectory)UsePatternMatching\CSharpAsAndNullCheckDiagnosticAnalyzer.Analyzer.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)UsePatternMatching\CSharpAsAndNullCheckDiagnosticAnalyzer.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)UsePatternMatching\CSharpIsAndCastCheckDiagnosticAnalyzer.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)UsePatternMatching\CSharpUseNotPattermDiagnosticAnalyzer.cs" />
```
Comment (sharwell) [inline src/Analyzers/CSharp/Analyzers/CSharpAnalyzers.projitems:99]:
```suggestion
    <Compile Include="$(MSBuildThisFileDirectory)UsePatternMatching\CSharpUseNotPatternDiagnosticAnalyzer.cs" />
```

Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Analyzers/CSharpAnalyzers.projitems:99]: Fixed

File: src/Analyzers/CSharp/CodeFixes/UsePatternMatching/CSharpUseNotPatternCodeFixProvider.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/CodeFixes/UsePatternMatching/CSharpUseNotPatternCodeFixProvider.cs:67]: tweaked things a bit.  we don't even run the analyzer now in code_style.

File: src/Analyzers/VisualBasic/Analyzers/UseIsNotExpression/VisualBasicUseIsNotDiagnosticAnalyzer.cs [hunk 7]
```diff
@@ -0,0 +1,82 @@
+﻿' Licensed to the .NET Foundation under one or more agreements.
```
Comment (davidwengier) [inline src/Analyzers/VisualBasic/Analyzers/UseIsNotExpression/VisualBasicUseIsNotDiagnosticAnalyzer.cs:1]: Cursed .cs file contains VB code. Much 2020, very meme.

Comment (CyrusNajmabadi) [reply src/Analyzers/VisualBasic/Analyzers/UseIsNotExpression/VisualBasicUseIsNotDiagnosticAnalyzer.cs]: apparently everything works fine (except automatic case completion).  i didn't even notice :)

**Pull request merged**